command/command.c

```diff
@@ -289,27 +304,126 @@ process_command_output (CommandApplet *command_applet, gchar *output)
     }
 }
 
+static gboolean
+stdout_io_func (GIOChannel *ioc, GIOCondition cond, gpointer data)
+{
+    CommandApplet *command_applet;
+
+    command_applet = data;
+    if (cond & (G_IO_IN | G_IO_PRI))
+    {
+        if (strlen (command_applet->buffer) == 0)
+        {
+            GError *error = NULL;
+            GIOStatus  ret;
+            gsize      len = 0;
+
+            ret = g_io_channel_read_chars (ioc, command_applet->buffer, command_applet->width, &len, &error);
+            if (len <= 0 || ret != G_IO_STATUS_NORMAL)
+            {
+                g_clear_error (&error);
+                return FALSE;
+            }
+            process_command_output (command_applet, command_applet->buffer);
+        }
+        return FALSE;
+    }
+
+    if (cond & (G_IO_ERR | G_IO_HUP | G_IO_NVAL))
+        return FALSE;
+
+    return TRUE;
+}
+
+static void
+on_child_exit (GPid child_pid, gint status, gpointer data)
+{
+    CommandApplet *command_applet;
+
+    command_applet = data;
+    g_spawn_close_pid (child_pid);
+    command_applet->child_pid = 0;
+}
+
+static GIOChannel *
+set_up_io_channel (gint fd, GIOCondition cond, GIOFunc func, gpointer data)
+{
+    GIOChannel *ioc;
+
+    ioc = g_io_channel_unix_new (fd);
+
+    g_io_channel_set_encoding (ioc, NULL, NULL);
+    g_io_channel_set_buffered (ioc, FALSE);
+
+    g_io_channel_set_close_on_unref (ioc, TRUE);
+
+    g_io_add_watch (ioc, cond, func, data);
+    g_io_channel_unref (ioc);
+
+    return ioc;
+}
+
 static gboolean
 command_execute (CommandApplet *command_applet)
 {
     GError *error = NULL;
-    gchar *output = NULL;
-    gint ret = 0;
+    gint argc;
+    gchar **argv;
+    gint stdout_fd;
 
-    if (g_spawn_command_line_sync (command_applet->command, &output, NULL, &ret, &error))
+    /* command is empty, wait for next timer execution */
+    if (strlen (command_applet->command) == 0)
     {
-        process_command_output (command_applet, output);
+        return TRUE;
     }
-    else
+
+    if (!g_shell_parse_argv (command_applet->command, &argc, &argv, &error))
```

Comment:
`argc` seems to be unused, it is optional parameter - you can pass `NULL` if you dont need it...